misk/src/main/kotlin/misk/web/actions/UpstreamResourceInterceptor.kt

```diff
@@ -76,31 +90,39 @@ class UpstreamResourceInterceptor(
   /**
    * Maps URLs requested against this server to URLs of servers to delegate to
    *
-   * localPathPrefix: `/_admin/`
-   * upstreamBaseUrl: `http://localhost:3000/`
+   * url_path_prefix: `/_admin/`
+   * server_url: `http://localhost:3000/`
    *
    * An incoming request then for `/_admin/config.js` would route to `http://localhost:3000/_admin/config.js`.
    */
+  // Make register tab a subclass of mapping
   data class Mapping(
-    val localPathPrefix: String,
-    val upstreamBaseUrl: HttpUrl
+    val urlPathPrefix: String,
+    val upstreamBaseUrl: HttpUrl,
+    val upstreamBaseJar: String,
+    val upstreamMode: Mode
   ) {
     init {
-      require(localPathPrefix.endsWith("/") &&
-          localPathPrefix.startsWith("/") &&
+      require(urlPathPrefix.startsWith("/") &&
+          urlPathPrefix.endsWith("/") &&
+          !urlPathPrefix.startsWith("/api/") &&
           upstreamBaseUrl.encodedPath().endsWith("/") &&
-          upstreamBaseUrl.pathSegments().size == 1)
+          upstreamBaseUrl.pathSegments().size == 1 &&
+          upstreamBaseJar.startsWith("/") &&
+          upstreamBaseJar.endsWith("/"))
     }
   }
 
+  enum class Mode {
+    SERVER,
+    JAR
```

Comment:
I have a different idea on how I’d like this to work.

I think what I want is that you either use a `UpstreamResourceInterceptor.Mapping` to get a resource from a URL, or you use something like a `ResourceInterceptor.Mapping` to get a resource from the JAR, and the caller always picks which to install.

Ie. this class doesn’t care whether you’re a SERVER or a JAR, it just does SERVER things. There’s another class that just does JAR things, and in the module you have to decide which you want.

Reply:
oh okay, yeah that makes more sense. I'll split out that functionality into a new `ResourceInterceptor` class then and add separate multibindings as required.

Reply:
Also, should this be called `JarResourceInterceptor` for better transparency of where it will be getting these resources. Or in Java ecosystem maybe that's a redundant prefix.

Reply:
How about this:

 * WebProxyInterceptor - what we’re currently calling UpstreamResourceInterceptor
 * ResourceInterceptor - serves from the .jar file or filesystem

Reply:
💯